### PR TITLE
Training Task# 14 Caching to improve the app performance

### PR DESCRIPTION
**Pull request by hamza-56**
- [x] Read about different cache backends (redis, memcached, file, database, local-memory cache (default))

- [x] Document steps to install memcache on mac [link](https://gist.github.com/hamza-56/8a25f21df95512f0627eb61a30edf05f)

- [x] Use memcached

- [x] Verify cache is working

Ticket link: https://edlyio.atlassian.net/browse/EDE-1731

### CODE REVIEW

**Comment by tasawernawaz** [comment]
@hamza-56 I can't see any usage of caching? 

One use case can be a list view, 
- check if data exists in the cache, if exists get it from the cache, if not get it from DB and store it in the cache.
- on insertion, you can invalidate the cache.

File: api/views.py [hunk 1]
```diff
@@ -17,6 +18,56 @@ class LogViewSet(viewsets.ModelViewSet):
...
+    def create(self, request):
+        response = super().create(request)
+        self._invalidate_list_cache()
+        return response
+
+    def update(self, request, pk=None):
+        cache_key = f'{self.DETAIL_CACHE_KEY_PREFIX}{pk}'
+
+        response = super().update(request, pk=pk)
+        cache.set(cache_key, response.data)
```
Comment (tasawernawaz) [inline api/views.py:59]: I believe this line won't have any impact? as we are invalidating cache on the next line.

Comment (tasawernawaz) [reply api/views.py:59]: also, we should invalidate the detail cache.

File: api/views.py [hunk 2]
```diff
@@ -17,6 +18,56 @@ class LogViewSet(viewsets.ModelViewSet):
...
+        self._invalidate_list_cache()
+        return response
+
+    def destroy(self, request, pk=None):
+        cache_key = f'{self.DETAIL_CACHE_KEY_PREFIX}{pk}'
+
+        response = super().destroy(request, pk=pk)
+        cache.set(cache_key, response.data)
+        self._invalidate_list_cache()
```
Comment (tasawernawaz) [inline api/views.py:68]: same as above

**Review by tasawernawaz — CHANGES_REQUESTED** [review]

File: api/views.py (hunk 1 above)
Comment (hamza-56) [reply api/views.py:59]: Line 59 updates the detail cache and next line invalidates the list cache

**Review by tasawernawaz — APPROVED** [review]

**Pull request merged**